plugins/modules/terraform.py

```diff
@@ -428,6 +436,42 @@ def build_plan(command, project_path, variables_args, state_file, targets, state
     ))
 
 
+def get_diff(diff_output):
+    def get_tf_resource_address(e):
+        return e['resource']
+
+    diff_json_output = json.loads(diff_output)
+    tf_reosource_changes = diff_json_output['resource_changes']
+    diff_after = []
+    diff_before = []
+    changed = False
+    for item in tf_reosource_changes:
+        item_change = item['change']
+        tf_before_state = {'resource': item['address'], 'change': item['change']['before']}
+        tf_after_state = {'resource': item['address'], 'change': item['change']['after']}
+
+        if item_change['actions'] == ['update'] or item_change['actions'] == ['delete', 'create']:
+            diff_before.append(tf_before_state)
+            diff_after.append(tf_after_state)
+            changed = True
+
+        if item_change['actions'] == ['delete']:
+            diff_before.append(tf_before_state)
+            changed = True
+
+        if item_change['actions'] == ['create']:
+            diff_after.append(tf_after_state)
+            changed = True
+
+    diff_before.sort(key=get_tf_resource_address)
+    diff_after.sort(key=get_tf_resource_address)
+
+    return changed, dict(
+        before=yaml.safe_dump(diff_before),
+        after=yaml.safe_dump(diff_after),
```

Comment:
Another question: why use YAML here and not JSON? Or (even better) simply returning the raw JSON data structure, so that the callback plugin configured by the user can format the data as it wants? (For that here objects need to be returned, not arrays; you can simply do `before={'data': diff_before}` for example.)

Reply:
In my opinion yaml output is more clear and readable than json.

I can handle both cases if yaml is exist the module return yaml and if it's not exist return json. but I have no idea how to mention that in documentations? @felixfontein

Reply:
Why not simply let the user decide what they prefer?

Reply:
I thought it's not common to ask about the format of the diff output.

BTW I have added `diff_in_yaml` option to show diff in yaml format PTAL

Reply:
Please remove that option. The right way to do this is by chosing which callback plugin you want to use, and by modules returning structured data and not pre-rendered text, and not by doing anything in the module.

For example, running some playbook with `default` callback (and default `json` output format):
```diff
TASK [community.general.docker_container] ****************************************************************************************************************************************************
--- before
+++ after
@@ -1,18 +1,19 @@
 {
     "command": [
-        "nginx",
-        "-g",
-        "daemon off;"
+        "/bin/sh",
+        "-c",
+        "sleep 10m"
     ],
     "env": [
-        "NGINX_VERSION=1.25.3",
-        "NJS_VERSION=0.8.2",
-        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
-        "PKG_RELEASE=1~bookworm"
+        "GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D",
+        "LANG=C.UTF-8",
+        "PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+        "PYTHON_GET_PIP_SHA256=96461deced5c2a487ddc65207ec5a9cffeca0d34e7af7ea1afc470ff0d746207",
+        "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/0d8570dc44796f4369b652222cf176b3db6ac70e/public/get-pip.py",
+        "PYTHON_PIP_VERSION=23.0.1",
+        "PYTHON_SETUPTOOLS_VERSION=65.5.1",
+        "PYTHON_VERSION=3.10.12"
     ],
-    "exposed_ports": [
-        "80/tcp",
-        "90/tcp"
-    ],
-    "image": "sha256:a8758716bb6aa4d90071160d27028fe4eaee7ce8166221a97d30440c8eac2be6"
+    "exposed_ports": [],
+    "image": "sha256:3898600d93b67fda2a9c6fdf189922444898516ded2040174c4bfc1129fefa31"
 }
```
Running the exact same playbook with `ANSIBLE_CALLBACK_RESULT_FORMAT=yaml` set:
```diff
TASK [community.general.docker_container] ****************************************************************************************************************************************************
--- before
+++ after
@@ -1,14 +1,16 @@
     command:
-    - nginx
-    - -g
-    - daemon off;
+    - /bin/sh
+    - -c
+    - sleep 10m
     env:
-    - NGINX_VERSION=1.25.3
-    - NJS_VERSION=0.8.2
-    - PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
-    - PKG_RELEASE=1~bookworm
-    exposed_ports:
-    - 80/tcp
-    - 90/tcp
-    image: sha256:a8758716bb6aa4d90071160d27028fe4eaee7ce8166221a97d30440c8eac2be6
+    - GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D
+    - LANG=C.UTF-8
+    - PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+    - PYTHON_GET_PIP_SHA256=96461deced5c2a487ddc65207ec5a9cffeca0d34e7af7ea1afc470ff0d746207
+    - PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/0d8570dc44796f4369b652222cf176b3db6ac70e/public/get-pip.py
+    - PYTHON_PIP_VERSION=23.0.1
+    - PYTHON_SETUPTOOLS_VERSION=65.5.1
+    - PYTHON_VERSION=3.10.12
+    exposed_ports: []
+    image: sha256:3898600d93b67fda2a9c6fdf189922444898516ded2040174c4bfc1129fefa31
```

For this the module doesn't need to do anything but return JSON objects for `before` and `after`.

Reply:
Thanks for your explanation, 

Done, PTALA!